### PR TITLE
sql: support all Kafka security protocols

### LINKED ISSUE #12401: Support for SASL_SSL / SCRAM-SHA-256
### What version of Materialize are you using?

 v0.26.1-dev (3fcd00c07)

### How did you install Materialize?

Materialize Cloud

### What is the issue?

I was using Upstash as a managed Kafka Broker.  The [documentation and integration](https://docs.upstash.com/kafka/howto/connect-with-clients) guides suggest using `SASL_SSL` as the security protocol and `SCRAM-SHA-256` as the sasl mechanisms.

I tried doing this with the following `CREATE SOURCE` command

```
materialize=> create source upstash
from kafka broker 'honest-prawn-11405-us1-kafka.upstash.io:9092' topic 'primary-data' WITH (
      security_protocol = 'SASL_SSL',
      sasl_mechanisms = 'SCRAM-SHA-256',
      sasl_username = <REMOVED>,
      sasl_password = <REMOVED> 
  )
FORMAT BYTES;
```

which fails with the following:

```
ERROR:  Failed to create and connect Kafka consumer: Client creation error: No provider for SASL mechanism SCRAM-SHA-256: recompile librdkafka with libsasl2 or openssl support. Current build options: PLAIN
```

Switching `sasl_mechanisms = 'PLAIN'` did not solve the problem either:

```
ERROR:  Failed to create and connect Kafka consumer: librdkafka: [thrd:sasl_ssl://honest-prawn-11405-us1-kafka.upstash.io:9092/bootstr]: sasl_ssl://honest-prawn-11405-us1-kafka.upstash.io:9092/bootstrap: SASL authentication error: Authentication failed: Invalid username or password (after 208ms in state AUTH_REQ)
```

but removing both `sasl_mechanisms` and `security_protocol` did work:

```
materialize=> create source upstash
from kafka broker 'honest-prawn-11405-us1-kafka.upstash.io:9092' topic 'primary-data' WITH (
      sasl_username = <REMOVED>,
      sasl_password = <REMOVED>
  )
FORMAT BYTES;
CREATE SOURCE
```

My issue is that `SASL_SSL` as the security protocol and `SCRAM-SHA-256` as the sasl mechanisms should work according to our documentation.

### Relevant log output

_No response_

### LINKED ISSUE #19039: Support `SASL_PLAINTEXT` authentication mechanism
### Feature request

Materialize Kafka only supports SASL_SSL. If a user tries to connect to brokers configured for SASL_PLAINTEXT we log an error and silently fail to ingest data. 

> sasl_ssl://20.100.194.205:9092/bootstrap: SSL handshake failed: Disconnected: connecting to a PLAINTEXT broker listener? (after 39ms in state SSL_HANDSHAKE

~Instead, we should expose an error via the source status table or at least issue a notice and fail to create the source~

*Nikhil edit:* We should just support `SECURITY PROTOCOL = SASL_PLAINTEXT`. We used to support this, pre-platform, but it got lost in the transition to connections.

### PR DESCRIPTION
**Pull request by benesch**
Through an accident of SQL parsing, we only supported the `SASL_SSL` and `SSL` Kafka security protocols, and the latter required TLS client authentication.

This commit teaches Materialize to also support:

  * `SECURITY PROTOCOL = PLAINTEXT`
  * `SECURITY PROTOCOL = SASL_PLAINTEXT`
  * `SECURITY PROTOCOL = SSL` without client authentication

This is, at last, the full matrix of Kafka security protocols.

Fix #12401.
Fix #19039.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

I didn't write a single test. Leaving in draft until 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Allow creating Kafka connections that use the `PLAINTEXT` and `SASL_PLAINTEXT` security protocols.
  - Allow Kafka connections to enable the `SSL` security protocol without enabling TLS client authentication—i.e., using TLS only for encryption.

### CODE REVIEW

**Comment by philip-stoev** [comment]
Ideally this required a test for each protocol. As that would be a lot of painful slogging, I have added it to the QA weekly agenda to discuss.

**Review by jkosh44 — APPROVED** [review]

**Comment by benesch** [comment]
> Ideally this required a test for each protocol. As that would be a lot of painful slogging, I have added it to the QA weekly agenda to discuss.

One thing that could make this easier: rather than having multiple different mzcompose compositions that each configure Kafka in a different way, and teaching testdrive how to talk to all those differently configured Kafkas, I think we want to instead run _one_ Kafka/CSR that has multiple different authentication methods configured. Imagine: a plaintext listener, a SSL listener with mTLS, an SSL listener without mTLS, a SASL_SSL listener, and a SASL_PLAINTEXT listener. Then we could have a _single_ testdrive file that could create a connection to e.g. kafka:9092 for the plaintext listener, kafka:9093 for the SSL listener, etc.

Also, testdrive's commands (e.g., `$ kafka-ingest`) could always talk to Kafka over the plaintext listener, so we wouldn't need to be plumbing a bunch of security options into testdrive. That would greatly simplify things.

**Review by guswynn — APPROVED** [review]
storage stuff looks good!

it may be easier to add an LD flag to control the default, and default to `PLAINTEXT` in mzcompose

**Review by def- — COMMENTED** [review]
Is it expected that this currently fails most existing tests with `ERROR: Broker does not support SSL connections`? https://buildkite.com/materialize/tests/builds/67875

**Comment by guswynn** [comment]
@def- yeah for now its changing the default for kafka connections to `SSL` over `PLAINTEXT`, so it will require some effort to pass ci (either a flag to control this or changing mzcomposes)

**Comment by def-** [comment]
I'll give that a try and try to add explicit tests too.

**Comment by benesch** [comment]
> I'll give that a try

I just pushed up a monster commit that adds `SECURITY PROTOCOL = PLAINTEXT` everywhere it's needed.

> and add explicit tests too.

I think our existing kafka-ssl and kafka-sasl-plain test suites need a good solid refactoring. There's too much incidental complexity in the way they configure the authentication parameters. I'm noodling on the refactor I described here: https://github.com/MaterializeInc/materialize/pull/22945#issuecomment-1795174018. Just pushed up what I have. I'm going to try to fill out a few more entries in the matrix, but after that I might need your help, @def-, to finish it out.

**Comment by benesch** [comment]
> > I'll give that a try
> 
> I just pushed up a monster commit that adds `SECURITY PROTOCOL = PLAINTEXT` everywhere it's needed.
> 
> > and add explicit tests too.
> 
> I think our existing kafka-ssl and kafka-sasl-plain test suites need a good solid refactoring. There's too much incidental complexity in the way they configure the authentication parameters. I'm noodling on the refactor I described here: [#22945 (comment)](https://github.com/MaterializeInc/materialize/pull/22945#issuecomment-1795174018). Just pushed up what I have. I'm going to try to fill out a few more entries in the matrix, but after that I might need your help, @def-, to finish it out.

There was no good pause point. The rewritten `kafka-auth` test suite is ready to go. I recommend reviewing it largely as new code: https://github.com/MaterializeInc/materialize/tree/d82b732d41aa1dc680c44b9515906611333359c7/test/kafka-auth

@jkosh44 @guswynn could you take a look at the migration I added to the latest revision? It adds an explicit `SECURITY PROTOCOL` to all existing Kafka connections to avoid the backward compatibility break. https://github.com/MaterializeInc/materialize/pull/22945/files#diff-8d2fe0db05338dd8df0efaffc2dbdeb4075ccf16e70f98ffa5911c75b0900974

**Comment by benesch** [comment]
Nightlies: https://buildkite.com/materialize/nightlies/builds/5152

**Review by def- — APPROVED** [review]
The testing part looks great to me! I was planning to switch all tests to use Kafka SSL connections instead of adding the `SECURITY PROTOCOL PLAINTEXT`, but I'm not sure if we want the extra complexity of that. I'll also check coverage (https://buildkite.com/materialize/coverage/builds/313) and try to write a platform-check to make sure the kafka connection options persist (doesn't have to be part of this PR).

**Comment by benesch** [comment]
> I was planning to switch all tests to use Kafka SSL connections instead of adding the SECURITY PROTOCOL PLAINTEXT, but I'm not sure if we want the extra complexity of that

Yeah, this seemed like the best option, painful as it was to get here:

  * It seems like not much overhead to type out `SECURITY PROTOCOL = PLAINTEXT` when you go to add a new Kafka test, and it's nice to be so explicit. (@guswynn suggested adding a LD flag that controlled the default, but whenever possible I like to minimize the difference between production and CI.)

  * There are at least a few engineers who still run testdrive outside of mzcompose (`bin/environmentd` in one terminal and `cargo run --bin testdrive` in another), where Kafka and the schema registry are run locally on the engineer's machine. It's hard enough to get Kafka/CSR running locally in plaintext mode; I'd hate to also make those folks set up TLS for their deployment. (It's also much harder to interact with a TLS-enabled Kafka cluster for tinkering, since you need to constantly be passing the CA into tools.)

File: test/kafka-auth/test-kafka-sasl-ssl.td [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+! CREATE CONNECTION kafka_invalid TO KAFKA (
+    BROKER 'kafka:9096',
+    SASL MECHANISMS 'GARBAGE',
+    SASL USERNAME 'materialize',
+    SASL PASSWORD SECRET password,
+    SSL CERTIFICATE AUTHORITY '${ca-crt}'
+  )
+contains:Unsupported SASL mechanism: GARBAGE
+
+! CREATE CONNECTION kafka_plain TO KAFKA (
```
Comment (def-) [inline test/kafka-auth/test-kafka-sasl-ssl.td:80]:
```suggestion
! CREATE CONNECTION kafka_invalid TO KAFKA (
```

**Comment by def-** [comment]
> Nightlies: https://buildkite.com/materialize/nightlies/builds/5152

Just saw that this has failures, I'll fix them up (and rebase the PR, and add my tests).

File: src/adapter/src/catalog/migrate.rs [hunk 2]
```diff
@@ -137,6 +141,65 @@ fn ast_rewrite_create_connection_options_0_77_0(
     Ok(())
 }
 
+/// Add `SECURITY PROTOCOL` to all existing Kafka connections, based on the
```
Comment (jkosh44) [inline src/adapter/src/catalog/migrate.rs:144]: I think you left off the second half of this sentence.

File: src/adapter/src/catalog/migrate.rs [hunk 3]
```diff
@@ -137,6 +141,65 @@ fn ast_rewrite_create_connection_options_0_77_0(
...
+                    "impossible Kafka connection with both SSL and SASL options: {:#?}",
+                    node.values
+                ),
+            };
+            node.values.push(ConnectionOption {
+                name: SecurityProtocol,
+                value: Some(WithOptionValue::Value(Value::String(
+                    security_protocol.into(),
+                ))),
+            });
```
Comment (jkosh44) [inline src/adapter/src/catalog/migrate.rs:195]: I have this whole PR paged out. Do we need to remove the existing values and replace them with this or adding a new value on top is fine?

File: test/kafka-auth/test-kafka-mssl.td [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+> CREATE SECRET kafka1_key AS '${kafka1-key}'
+> CREATE SECRET garbage_key AS 'garbage'
+
+$ kafka-create-topic topic=text-data
+$ kafka-ingest topic=text-data format=bytes
+banana
+
+# Test invalid configurations.
+
+# This is a really bad error message to indicate "client certificate required",
```
Comment (def-) [inline test/kafka-auth/test-kafka-mssl.td:28]: By upgrading Kafka the error became more reasonable.

**Review by def- — COMMENTED** [review]
Tests & nightly are good now, Most of nightly was green in https://buildkite.com/materialize/nightlies/builds/5163
The rest was green in https://buildkite.com/materialize/nightlies/builds/5169
Feature-benchmark is not possible to make green in the PR (since it runs against main, and main has the same version as this PR), but should be green on main since it runs against the released version (v0.77) there.

File: src/adapter/src/catalog/migrate.rs (hunk 3 above)
Comment (guswynn) [reply src/adapter/src/catalog/migrate.rs:195]: Its fine because `SecurityProtocol` does not exist in older versions of mz

**Review by guswynn — APPROVED** [review]
migration looks good, and what a lift to fix all the td's

**Review by jkosh44 — APPROVED** [review]
Migration LGTM, other than the doc-comment being incomplete.

**Comment by benesch** [comment]
> Migration LGTM, other than the doc-comment being incomplete.

Will fix in a follow-up commit. Want to get this in while it's green. @def-: thank you SO much for all the nightly fixes!

**Pull request merged**